### PR TITLE
Charge using _cost in reducer

### PR DESCRIPTION
**Pull request by lukasz-golebiewski**
## Overview
Replaces the CostAccounting in Reducer with the _cost alias

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-2985

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by lukasz-golebiewski** [comment]
bors try

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 1]
```diff
@@ -19,7 +19,7 @@ trait Substitute[M[_], A] {
 }
 
 object Substitute {
-  def charge[A: Chargeable, M[_]: Sync: CostAccounting](
+  private[interpreter] def charge[A: Chargeable, M[_]: Sync: CostAccounting](
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: I think making this private could be a separate commit :)

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 2]
```diff
@@ -1390,7 +1395,7 @@ object Reduce {
           toArgRaw   <- evalToLong(args(1))
           toArg      <- restrictToInt(toArgRaw)
           result     <- slice(baseExpr, fromArg, toArg)
-          _          <- costAccountingAlg.charge(sliceCost(toArg))
+          _          <- charge[M](sliceCost(toArg))
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: AFAIK you shouldn't need the `[M]` type annotation in calls to `charge`.

Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: scalac disagrees

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 1 above)
Comment (lukasz-golebiewski) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Done

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala [hunk 3]
```diff
@@ -1,3 +1,23 @@
...
-package object accounting extends Costs
+import cats.Monad
+import cats.implicits._
+import cats.mtl.MonadState
+
+import coop.rchain.rholang.interpreter.errors.OutOfPhlogistonsError
+
+package object accounting extends Costs {
+
+  type _cost[F[_]] = MonadState[F, Cost]
```
Comment (rabbitonweb) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: consider using `Cell` (with `flatModify`, without `set`) in your next PR

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: I really don't get why if MonadState suffices. `set` is a direct consequence of `modify`, which cell also has. Can you elaborate on the reasons for using Cell?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala [hunk 4]
```diff
@@ -15,10 +15,9 @@ package object accounting extends Costs {
       amount: Cost
   )(implicit cost: _cost[F], error: _error[F]): F[Unit] =
     for {
-      currentCost <- cost.get
-      newCost     = currentCost - amount
-      _           <- cost.set(newCost)
-      _           <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
+      _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
+      _ <- cost.modify(_ - amount)
+      _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
```
Comment (ArturGajowy) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: :kissing_heart: :ok_hand:

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala [hunk 5]
```diff
@@ -15,10 +15,9 @@ package object accounting extends Costs {
       amount: Cost
   )(implicit cost: _cost[F], error: _error[F]): F[Unit] =
     for {
-      currentCost <- cost.get
-      newCost     = currentCost - amount
-      _           <- cost.set(newCost)
-      _           <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
+      _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
```
Comment (goral09) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: I am not sure that double check is necessary. If every time the phlos available is updated with this method then if the limit goes below 0 it will break during exit.

Comment (ArturGajowy) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala]: good point! Having just the check immediately after the `modify` should suffice :+1:
